src/hydra-queue-runner/metrics.hh

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#include <cassert>
+#include <prometheus/counter.h>
+#include <prometheus/gauge.h>
+#include <prometheus/histogram.h>
+#include <prometheus/registry.h>
+#include "sync.hh"
+
+// Track the elapsed time of a code block, and manually submit the timing
+// data to a prometheus Histogram.
+struct PromTimerManual
+{
+    std::chrono::time_point<std::chrono::high_resolution_clock> created;
+
+    PromTimerManual()
+        : created(std::chrono::high_resolution_clock::now())
+    {
+
+    }
+
+public:
+    void finish(prometheus::Histogram& histogram)
+    {
+        histogram.Observe(std::chrono::duration<float>(std::chrono::high_resolution_clock::now() - created).count());
+    }
+};
+
+// Track the elapsed time of a code block, automatically submitting the timingh
+// data when the timer drops.
+struct PromTimer
+{
+    PromTimerManual timer;
+    prometheus::Histogram& histogram;
+
+    PromTimer(prometheus::Histogram& metric)
+        : histogram(metric)
+    {
+
+    }
+
+    ~PromTimer() {
+        timer.finish(histogram);
+    }
+};
+
+// Track the elapsed time of a code block where multiple exits are possible.
+//
+// The timing will be submitted to the constructor's histogram if the timer
+// drops before finish() was called manually.
+//
+// Complicated code paths should explicitly finish() every exit, leaving the
+// default histogram as a way to track failures to properly track exits.
+struct PromTimerExactlyOneExit
+{
+    PromTimerManual timer;
+    prometheus::Histogram& histogram;
+    bool unsubmitted;
+
+    PromTimerExactlyOneExit(prometheus::Histogram& default_metric)
+        : histogram(default_metric)
+        , unsubmitted(true)
+    {
+
+    }
+
+    ~PromTimerExactlyOneExit() {
+        if (unsubmitted) {
+            finish(histogram);
+        }
+    }
+
+public:
+    void finish(prometheus::Histogram& metric)
+    {
+        // Assert that we have not reported any metric yet,
+        // as this timer should only be finish()'d once
+        assert(unsubmitted);
```

Comment:
It doesn't seem right to assert here. Log a warning or make it only assert while you execute the unit tests. Usually you don't want to kill the process because some of the metrics arent "correct". Especially given how sad the unit test story is.

Reply:
If this is a logic error that shouldn't happen, then an assertion is fine.